### PR TITLE
lib: use ext helper in more places, test custom exts.

### PR DESCRIPTION
**Pull request by cpu**
This branch updates a couple places that were writing x509 extensions by hand to use the existing `write_x509_extension` helper instead. 

Since custom extensions are a bit trickier a unit test for custom extensions in CSRs and certificates is added before refactoring the existing code to use the helper. This helps ensure the change is a no-op from the perspective of users.

With the updated test, it was possible to demonstrate a bug similar to https://github.com/rustls/rcgen/issues/122. The CSR serialization logic requiring SANs to be present in order to emit the custom extensions into the CSR PKCS9 extension request attribute. This branch fixes the issue.

### CODE REVIEW

**Review by est31 — APPROVED** [review]
I only gave this a cursory glance due to limited time but the general idea of the PR is good, thanks. Maybe it would be good to get another review?

**Comment by cpu** [comment]
> Maybe it would be good to get another review?

Sounds good. @djc when you have some time I'd appreciate a review pass.

FWIW I'm also working on a subsequent refactoring of extension handling to fix #122 - I'll open that as a follow-up shortly.

**Comment by cpu** [comment]
> FWIW I'm also working on a subsequent refactoring of extension handling to fix https://github.com/rustls/rcgen/issues/122 - I'll open that as a follow-up shortly.

General idea is in https://github.com/cpu/rcgen/pull/1 I haven't done a self-review pass yet and I'd like to add more tests.

File: src/lib.rs [hunk 1]
```diff
@@ -911,7 +911,7 @@ impl CertificateParams {
 			// Write extensions
 			// According to the spec in RFC 2986, even if attributes are empty we need the empty attribute tag
 			writer.next().write_tagged(Tag::context(0), |writer| {
-				if !subject_alt_names.is_empty() {
+				if !subject_alt_names.is_empty() || !custom_extensions.is_empty() {
```
Comment (djc) [inline src/lib.rs:914]: Given the commit message text about a test failure, should this commit be moved around to appear before some of the other commits in this PR?

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Nice!

File: src/lib.rs (hunk 1 above)
Comment (cpu) [reply src/lib.rs:914]: I made the test change + fix in the same commit so they would all build green. The commit message text describes a manually created intermediate step.

**Pull request merged**